### PR TITLE
Yet Another Trusted Device PR

### PR DESCRIPTION
**Pull request by lizduty**
Basically just rebased #1027 and fixed the tests

### CODE REVIEW

File: pkg/provider/okta/okta.go [hunk 1]
```diff
@@ -1236,3 +1381,214 @@ func fidoWebAuthn(oc *Client, oktaOrgHost string, challengeContext *mfaChallenge
 
 	return gjson.GetBytes(body, "sessionToken").String(), nil
 }
+
+func verifyTrustedCert(oc *Client, doc *goquery.Document, duoHost string, duoSubmitURL string, q url.Values) (*goquery.Document, error) {
```
Comment (lizduty) [inline pkg/provider/okta/okta.go:1385]: this is the same as the existing PR, just broken out into helper functions to make testing easier

File: pkg/provider/okta/okta_test.go [hunk 2]
```diff
@@ -355,3 +356,191 @@ func TestGetStateToken(t *testing.T) {
...
+
+	host = upstreamURL.Host
+	submitURL := ts.URL + "/submit"
+
+	q := url.Values{}
+	q.Add("tx", "1234567890")
+	q.Add("parent", fmt.Sprintf("https://%s/signin/verify/duo/web", host))
+	q.Add("v", "2.8")
+
+	fakeForm := fmt.Sprintf(`
```
Comment (lizduty) [inline pkg/provider/okta/okta_test.go:451]: originally this was a constant, but it needs the http server address and it's only available at runtime

File: pkg/provider/okta/okta_test.go [hunk 3]
```diff
@@ -355,3 +356,191 @@ func TestGetStateToken(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, "token1", stateToken)
 }
+
+// anonymised from actual endpoint
+const fakeEndpointForm = `
+<form method="post" id="endpoint-health-form">
```
Comment (lizduty) [inline pkg/provider/okta/okta_test.go:362]: I copied this out of dev tools and changed all the PII, so it's reasonably accurate

File: pkg/provider/okta/okta.go [hunk 4]
```diff
@@ -1236,3 +1381,214 @@ func fidoWebAuthn(oc *Client, oktaOrgHost string, challengeContext *mfaChallenge
...
+
+	txid, _ := doc.Find("input[name=\"txid\"]").Attr("value")
+	sid, _ := doc.Find("input[name=\"sid\"]").Attr("value")
+	ehServiceUrl, _ := doc.Find("input[name=\"eh_service_url\"]").Attr("value")
+	akey, _ := doc.Find("input[name=\"akey\"]").Attr("value")
+	responseTimeout, _ := doc.Find("input[name=\"response_timeout\"]").Attr("value")
+	parent, _ := doc.Find("input[name=\"parent\"]").Attr("value")
+	duoAppUrl, _ := doc.Find("input[name=\"duo_app_url\"]").Attr("value")
+	ehDownloadLink, _ := doc.Find("input[name=\"eh_download_link\"]").Attr("value")
+	// isSilentCollection, _ := doc.Find("input[name=\"is_silent_collection\"]").Attr("value")
```
Comment (lizduty) [inline pkg/provider/okta/okta.go:1494]: this wasn't present in my debug session and seems to not be necessary

**Review by mapkon — CHANGES_REQUESTED** [review]
Please add tests to this. Thanks

**Comment by lizduty** [comment]
> Please add tests to this. Thanks

there's tests for the two helper functions introduced. Did you mean for the U2F flow as well?

**Comment by lizduty** [comment]
@mapkon please let me know if there's any other tests you'd like

**Comment by mapkon** [comment]
Thanks @lizduty I think we should wait for the code-coverage report.

@gliptak Do you know why it is not showing up on this?

File: pkg/provider/okta/okta.go (hunk 4 above)
Comment (mapkon) [reply pkg/provider/okta/okta.go:1494]: Can you remove it it? Commented out code is dead code

**Comment by mapkon** [comment]
@gliptak Got it. Please ignore my previous comment

**Review by mapkon — APPROVED** [review]

**Pull request merged**